src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs

```diff
@@ -0,0 +1,415 @@
+                    Dictionary<string, int> vocab,
+                    IReadOnlyDictionary<string, int>? specialTokensEncoder = null,
+                    int cacheSize = LruCache<string, int[]>.DefaultCacheSize) : this(cacheSize)
+        {
+            Debug.Assert(encoder is not null);
+            Debug.Assert(decoder is not null);
+            Debug.Assert(vocab is not null);
+
+            _encoder = encoder!;
+            _vocab = vocab!;
+            _decoder = decoder!;
+
+            _specialTokensEncoder = specialTokensEncoder;
+            if (_specialTokensEncoder is not null)
+            {
+                _specialTokensDecoder = _specialTokensEncoder.ToDictionary(kvp => kvp.Value, kvp => kvp.Key);
+            }
+        }
+
+        private Tiktoken(int cacheSize)
+        {
+            _cache = new LruCache<string, int[]>(cacheSize);
+        }
+
+        private void Initialize(Stream tikTokenBpeFileStream, IReadOnlyDictionary<string, int>? specialTokensEncoder = null)
+        {
+            if (tikTokenBpeFileStream is null)
+            {
+                throw new ArgumentNullException(nameof(tikTokenBpeFileStream));
+            }
+
+            (_encoder, _vocab, _decoder) = LoadTikTokenBpe(tikTokenBpeFileStream);
+
+            _specialTokensEncoder = specialTokensEncoder;
+            if (_specialTokensEncoder is not null)
+            {
+                _specialTokensDecoder = _specialTokensEncoder.ToDictionary(kvp => kvp.Value, kvp => kvp.Key);
+            }
+        }
+
+        /// <summary>
+        /// Load BPE rank dictionary from a stream.
+        /// </summary>
+        /// <param name="tikTokenBpeFileStream">Stream to the BPE rank file</param>
+        /// <returns>Map of byte[] to integer token id</returns>
+        /// <exception cref="InvalidOperationException"></exception>
+        internal static (Dictionary<byte[], int>, Dictionary<string, int>, IReadOnlyDictionary<int, byte[]>) LoadTikTokenBpe(Stream tikTokenBpeFileStream)
+        {
+            var encoder = new Dictionary<byte[], int>(new ByteArrayComparer());
+            var vocab = new Dictionary<string, int>();
+            var decoder = new Dictionary<int, byte[]>();
+
+            try
+            {
+                using (StreamReader reader = new StreamReader(tikTokenBpeFileStream))
+                {
+                    while (!reader.EndOfStream)
+                    {
+                        string line = reader.ReadLine();
+                        if (string.IsNullOrWhiteSpace(line))
+                        {
+                            continue;
+                        }
+
+                        var tokens = line.Split(' ');
+                        if (tokens.Length != 2)
+                        {
+                            throw new FormatException($"Invalid format in the BPE encoder file stream");
+                        }
+
+                        byte[] tokenBytes = Convert.FromBase64String(tokens[0]);
```

Comment:
I realize this is only done once per Tiktoken object, but it'd be nice to reduce the allocation / overhead involved here, e.g. not using split on each line.